src/SignalR/common/Shared/AsyncEnumerableAdapters.cs

```diff
@@ -13,38 +13,63 @@ namespace Microsoft.AspNetCore.SignalR.Internal
     // True-internal because this is a weird and tricky class to use :)
     internal static class AsyncEnumerableAdapters
     {
-        public static IAsyncEnumerable<object?> MakeCancelableAsyncEnumerable<T>(IAsyncEnumerable<T> asyncEnumerable, CancellationToken cancellationToken = default)
+        public static IAsyncEnumerator<object?> MakeCancelableAsyncEnumerator<T>(IAsyncEnumerable<T> asyncEnumerable, CancellationToken cancellationToken = default)
         {
-            return new CancelableAsyncEnumerable<T>(asyncEnumerable, cancellationToken);
+            var enumerator = asyncEnumerable.GetAsyncEnumerator(cancellationToken);
+            return enumerator as IAsyncEnumerator<object?> ?? new BoxedAsyncEnumerator<T>(enumerator);
         }
 
         public static IAsyncEnumerable<T> MakeCancelableTypedAsyncEnumerable<T>(IAsyncEnumerable<T> asyncEnumerable, CancellationTokenSource cts)
         {
             return new CancelableTypedAsyncEnumerable<T>(asyncEnumerable, cts);
         }
 
-#if NETCOREAPP
-        public static async IAsyncEnumerable<object?> MakeAsyncEnumerableFromChannel<T>(ChannelReader<T> channel, [EnumeratorCancellation] CancellationToken cancellationToken = default)
+        public static IAsyncEnumerator<object?> MakeAsyncEnumeratorFromChannel<T>(ChannelReader<T> channel, CancellationToken cancellationToken = default)
         {
-            await foreach (var item in channel.ReadAllAsync(cancellationToken))
-            {
-                yield return item;
-            }
+            return new ChannelAsyncEnumerator<T>(channel, cancellationToken);
         }
-#else
-        // System.Threading.Channels.ReadAllAsync() is not available on netstandard2.0 and netstandard2.1
-        // But this is the exact same code that it uses
-        public static async IAsyncEnumerable<object?> MakeAsyncEnumerableFromChannel<T>(ChannelReader<T> channel, [EnumeratorCancellation] CancellationToken cancellationToken = default)
+
+        private class ChannelAsyncEnumerator<T> : IAsyncEnumerator<object?>
         {
-            while (await channel.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
+            private readonly ChannelReader<T> _channel;
+            private readonly CancellationToken _cancellationToken;
+            public object? Current { get; private set; }
```

Comment:
Nit: we usually put properties below constructors and above methods.

Reply:
Ack, I follow up on all comments added. Thanks for the review.